In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import os

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [3]:
df=pd.read_csv("D:\ipl\IPL.csv")

C:\Users\karan\AppData\Local\Temp\ipykernel_20464\4278241017.py:1: DtypeWarning: Columns (28,29,30,31,43,46,47,48,51) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv("D:\ipl\IPL.csv")


In [4]:
df.shape

(278205, 64)

In [5]:
df.tail(5)

,Unnamed: 0,match_id,date,match_type,event_name,innings,batting_team,bowling_team,over,ball,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
278200,277951,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,2,19.2,Shashank Singh,6,0,1,JR Hazlewood,1,0,0,0,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,162,116,7,NaN,39,26,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278201,277952,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,3,19.3,Shashank Singh,6,6,1,JR Hazlewood,1,0,6,6,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,168,117,7,NaN,45,27,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278202,277953,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,4,19.4,Shashank Singh,6,4,1,JR Hazlewood,1,0,4,4,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,172,118,7,NaN,49,28,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278203,277954,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,5,19.5,Shashank Singh,6,6,1,JR Hazlewood,1,0,6,6,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,178,119,7,NaN,55,29,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278204,277955,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,6,19.6,Shashank Singh,6,6,1,JR Hazlewood,1,0,6,6,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,184,120,7,NaN,61,30,0,"('KA Jamieson', 'Shashank Singh')",NaN,False


In [6]:
df['match_id'].nunique()

1169

1. the data is of 1169 ipl matches from 2008-2025
2. average matches per season is 69

In [7]:
df.isnull().sum()

Unnamed: 0                0
match_id                  0
date                      0
match_type                0
event_name                0
innings                   0
batting_team              0
bowling_team              0
over                      0
ball                      0
ball_no                   0
batter                    0
bat_pos                   0
runs_batter               0
balls_faced               0
bowler                    0
valid_ball                0
runs_extras               0
runs_total                0
runs_bowler               0
runs_not_boundary         0
extra_type           263072
non_striker               0
non_striker_pos           0
wicket_kind          264382
player_out           264382
fielders             268192
runs_target          144302
review_batter        277333
team_reviewed        277333
review_decision      277333
umpire               277333
umpires_call              0
player_of_match           0
match_won_by              0
win_outcome         

there are plenty of missing value but most of them are maybe due to data error .we can resolve that using the map function

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278205 entries, 0 to 278204
Data columns (total 64 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         278205 non-null  int64  
 1   match_id           278205 non-null  int64  
 2   date               278205 non-null  object 
 3   match_type         278205 non-null  object 
 4   event_name         278205 non-null  object 
 5   innings            278205 non-null  int64  
 6   batting_team       278205 non-null  object 
 7   bowling_team       278205 non-null  object 
 8   over               278205 non-null  int64  
 9   ball               278205 non-null  int64  
 10  ball_no            278205 non-null  float64
 11  batter             278205 non-null  object 
 12  bat_pos            278205 non-null  int64  
 13  runs_batter        278205 non-null  int64  
 14  balls_faced        278205 non-null  int64  
 15  bowler             278205 non-null  object 
 16  va

In [9]:
df.duplicated().sum()

np.int64(0)

# data cleaning

In [10]:
train_df=df.copy()

In [11]:
train_df.tail()

,Unnamed: 0,match_id,date,match_type,event_name,innings,batting_team,bowling_team,over,ball,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
278200,277951,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,2,19.2,Shashank Singh,6,0,1,JR Hazlewood,1,0,0,0,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,162,116,7,NaN,39,26,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278201,277952,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,3,19.3,Shashank Singh,6,6,1,JR Hazlewood,1,0,6,6,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,168,117,7,NaN,45,27,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278202,277953,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,4,19.4,Shashank Singh,6,4,1,JR Hazlewood,1,0,4,4,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,172,118,7,NaN,49,28,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278203,277954,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,5,19.5,Shashank Singh,6,6,1,JR Hazlewood,1,0,6,6,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,178,119,7,NaN,55,29,0,"('KA Jamieson', 'Shashank Singh')",NaN,False
278204,277955,1473511,2025-06-03,T20,Indian Premier League,2,Punjab Kings,Royal Challengers Bengaluru,19,6,19.6,Shashank Singh,6,6,1,JR Hazlewood,1,0,6,6,False,NaN,KA Jamieson,9,NaN,NaN,NaN,191.0,NaN,NaN,NaN,NaN,False,KH Pandya,Royal Challengers Bengaluru,6 runs,Punjab Kings,field,"Narendra Modi Stadium, Ahmedabad",Ahmedabad,3,6,2025,2025,male,club,NaN,NaN,NaN,6,20,Unknown,Final,Unknown,184,120,7,NaN,61,30,0,"('KA Jamieson', 'Shashank Singh')",NaN,False


droping the unnamed :0 ,match type, event name : as these are same for all the mathes.and  ipl are of 20 overs

In [12]:
train_df.drop(columns=['Unnamed: 0','match_type','event_name'],inplace=True)

In [13]:
train_df.head()

,match_id,date,innings,batting_team,bowling_team,over,ball,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,2008-04-18,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,0.1,SC Ganguly,1,0,1,P Kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,2008-04-18,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,0.2,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,2008-04-18,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,0.3,BB McCullum,2,0,0,P Kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,2008-04-18,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,0.3,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,2008-04-18,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,0.4,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


we are also deleting the date column as it is already extracted into day /month /year

In [14]:
train_df.drop('date',axis=1,inplace=True)

In [15]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,0.1,SC Ganguly,1,0,1,P Kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,0.2,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,0.3,BB McCullum,2,0,0,P Kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,0.3,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,0.4,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


inning is defining the which club is playing first/bating  and which club is bowling  

replacing the names of batting team into the sort form (example "Kolkata Knight Riders"-kkr) and so on 

In [16]:
train_df['batting_team'].unique()

array(['Kolkata Knight Riders', 'Royal Challengers Bangalore',
       'Chennai Super Kings', 'Kings XI Punjab', 'Rajasthan Royals',
       'Delhi Daredevils', 'Mumbai Indians', 'Deccan Chargers',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Rising Pune Supergiants', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Lucknow Super Giants', 'Gujarat Titans',
       'Royal Challengers Bengaluru'], dtype=object)

some of the teams are have 2 value so we will convert these into lower case teams

In [17]:
train_df['batting_team'].nunique()

19

In [18]:
batting_dic={
            'Kolkata Knight Riders':'KKR',
            'Royal Challengers Bangalore':'RCB',
            'Chennai Super Kings':'CSK',
            'Kings XI Punjab':'KXP',
            'Rajasthan Royals':'RR',
            'Delhi Daredevils':'DC',
            'Mumbai Indians':'MI',
            'Deccan Chargers':'DEC',
            'Kochi Tuskers Kerala':'KTK',
            'Pune Warriors':"PW",
            'Sunrisers Hyderabad':'SRH',
            'Rising Pune Supergiants':'RPS',
            'Gujarat Lions':'GL',
            'Rising Pune Supergiant':'RPS',
            'Delhi Capitals':'DC',
            'Punjab Kings':'KXP',
            'Lucknow Super Giants':'LSG',
            'Gujarat Titans':'GT',
            'Royal Challengers Bengaluru':'RCB'}

In [19]:
# train_df.apply(lambda x : x .replace(batting_dic))
# this process can be slow so insted of using lamda we are going to use the replace function of pandas which is faster or the map function 
# Create a list of columns that contain team names
team_cols = ['batting_team', 'bowling_team'] 

# Option A: Using replace (Great for whole DataFrames or specific columns)
# Strip spaces and replace names for both columns efficiently
for col in team_cols:
    train_df[col] = train_df[col].str.strip().replace(batting_dic)# we added the strip fuction so that the there is no extrea space in
# the daataset which can cause problem in the future

In [20]:
train_df.head(8)

,match_id,innings,batting_team,bowling_team,over,ball,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,1,0.1,SC Ganguly,1,0,1,P Kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,2,0.2,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,3,0.3,BB McCullum,2,0,0,P Kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,3,0.3,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,4,0.4,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False
5,335982,1,KKR,RCB,0,5,0.5,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,5,0,NaN,0,4,0,"('BB McCullum', 'SC Ganguly')",NaN,False
6,335982,1,KKR,RCB,0,6,0.6,BB McCullum,2,0,1,P Kumar,1,1,1,0,False,legbyes,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,3,6,0,NaN,0,5,0,"('BB McCullum', 'SC Ganguly')",NaN,False
7,335982,1,KKR,RCB,1,1,1.1,BB McCullum,2,0,1,Z Khan,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,3,7,0,NaN,0,6,0,"('BB McCullum', 'SC Ganguly')",NaN,False


In [21]:
train_df[['match_id','ball','ball_no']].head(20)

,match_id,ball,ball_no
0,335982,1,0.1
1,335982,2,0.2
2,335982,3,0.3
3,335982,3,0.3
4,335982,4,0.4
5,335982,5,0.5
6,335982,6,0.6
7,335982,1,1.1
8,335982,2,1.2
9,335982,3,1.3


we dont need both ball and ball no so we are removingthe ball as the ball_no is same but better than ball

In [22]:
train_df.drop('ball',axis=1,inplace=True)

In [23]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,SC Ganguly,1,0,1,P Kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,BB McCullum,2,0,0,P Kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,BB McCullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


bat_pos define the order on which the batsman have enter the ground and contains 1-11 number a


In [24]:
train_df['bat_pos'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [25]:
# Group the data as you did
pos_stats = train_df.groupby(['batter', 'bat_pos'])['runs_batter'].sum().reset_index()

# Filter: Only show positions where the player has scored at least 30 runs
# This keeps your charts from looking cluttered with '1' or '0' run bars
pos_stats = pos_stats[pos_stats['runs_batter'] > 30]

In [26]:
pos_stats.head(10)

,batter,bat_pos,runs_batter
2,A Ashish Reddy,7,141
3,A Ashish Reddy,8,85
4,A Ashish Reddy,9,33
6,A Badoni,3,44
7,A Badoni,4,58
8,A Badoni,5,285
9,A Badoni,6,340
10,A Badoni,7,232
14,A Chopra,2,38
21,A Flintoff,5,38


In [27]:
train_df['batter']=train_df['batter'].apply(lambda x : x.lower())

runs_batter define the run made by the batsman in matches 

In [28]:
train_df.groupby('batter')['runs_batter'].sum().sort_values(ascending=False).tail(500)

batter
y nagar                    285
gh vihari                  284
vvs laxman                 282
b chipli                   280
a ashish reddy             280
mj lumb                    278
jp inglis                  278
pbb rajapaksa              277
pk garg                    273
mj guptill                 271
hv patel                   270
r sathish                  270
jc archer                  262
atharva taide              260
m kaif                     259
jo holder                  259
v suryavanshi              252
k gowtham                  247
harpreet brar              244
sm katich                  241
a mhatre                   240
bcj cutting                238
md mishra                  237
aniket verma               236
mohammad nabi              221
k goel                     218
ramandeep singh            217
aa jhunjhunwala            217
k rabada                   215
r dhawan                   210
ut yadav                   208
ca ingram                  205
p

In [29]:
wide_balls=train_df[train_df['balls_faced']==0]

In [30]:
wide_balls['extra_type'].value_counts()

extra_type
wides             9079
penalty, wides       1
Name: count, dtype: int64

In [31]:
wide_balls[wide_balls['extra_type']!='wides'].head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
98231,729309,2,DC,MI,4,4.2,m vijay,2,0,0,SL Malinga,0,7,7,2,False,"penalty, wides",Q de Kock,1,NaN,NaN,NaN,126.0,NaN,NaN,NaN,NaN,False,M Vijay,Delhi Daredevils,6 wickets,Mumbai Indians,bat,Sharjah Cricket Stadium,Unknown,27,4,2014,2014,male,club,NaN,NaN,NaN,6,20,16,Unknown,Unknown,30,25,0,NaN,8,12,0,"('M Vijay', 'Q de Kock')",NaN,False


there is only 1 penalty ball so we can say the balled_faced into wide balls 

In [32]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,P Kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,P Kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


ball faced 

In [33]:
train_df['balls_faced'].nunique()

2

In [34]:
train_df.groupby('batter')['balls_faced'].sum().sort_values(ascending=False).head(10)

batter
v kohli       6523
rg sharma     5337
s dhawan      5326
da warner     4702
sk raina      4046
am rahane     4025
ms dhoni      3957
kl rahul      3848
rv uthappa    3801
kd karthik    3580
Name: balls_faced, dtype: int64

In [35]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,P Kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,P Kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,P Kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


In [36]:
train_df['bowler'].unique()

array(['P Kumar', 'Z Khan', 'AA Noffke', 'JH Kallis', 'SB Joshi',
       'CL White', 'AB Dinda', 'I Sharma', 'AB Agarkar', 'SC Ganguly',
       'LR Shukla', 'B Lee', 'S Sreesanth', 'JR Hopes', 'IK Pathan',
       'K Goel', 'PP Chawla', 'WA Mota', 'JDP Oram', 'MS Gony',
       'M Muralitharan', 'P Amarnath', 'Joginder Sharma', 'GD McGrath',
       'B Geeves', 'MF Maharoof', 'R Bhatia', 'DL Vettori', 'MM Patel',
       'SR Watson', 'SK Trivedi', 'SK Warne', 'YK Pathan', 'D Salunkhe',
       'R Vinay Kumar', 'B Akhil', 'A Nehra', 'SM Pollock', 'DS Kulkarni',
       'ST Jayasuriya', 'Harbhajan Singh', 'AM Nayar', 'M Kartik',
       'Mohammad Hafeez', 'DJ Hussey', 'WPUJC Vaas', 'RP Singh',
       'SB Styris', 'SB Bangar', 'A Symonds', 'PP Ojha', 'Pankaj Singh',
       'Mohammad Asif', 'VY Mahesh', 'Shahid Afridi', 'DJ Bravo',
       'VS Yeligati', 'MA Khote', 'D Kalyankrishna', 'VRV Singh',
       'Sohail Tanvir', 'A Kumble', 'DNT Zoysa', 'SD Chitnis',
       'Yuvraj Singh', 'Shoaib Malik',

In [37]:
train_df['bowler']=train_df['bowler'].apply(lambda x:x.lower())

In [38]:
train_df.groupby('bowler')['valid_ball'].value_counts().reset_index().sort_values(by='count',ascending=False).head(10)

,bowler,valid_ball,count
679,r ashwin,1,4710
872,sp narine,1,4351
132,b kumar,1,4222
712,ra jadeja,1,4056
664,pp chawla,1,3850
1001,ys chahal,1,3791
296,harbhajan singh,1,3416
13,a mishra,1,3371
370,jj bumrah,1,3359
98,ar patel,1,3321


here the bowler defines the name of the bowler[50] total number of bowler and weather the balll is valid or not 

In [39]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,p kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,p kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


In [40]:
train_df['runs_extras'].unique()

array([1, 0, 5, 4, 2, 3, 7])

DEFINES THE EXTRA RUNS A BOLLWER GETS

In [41]:
train_df.groupby('bowler')['runs_extras'].sum().sort_values(ascending=False).head(10)

bowler
b kumar           320
sl malinga        292
r ashwin          263
mohammed siraj    261
dj bravo          258
ut yadav          257
i sharma          247
hv patel          243
jj bumrah         238
p kumar           236
Name: runs_extras, dtype: int64

runs_extra defines the extra run which are faced by the bowler inclide[legby,wide,freehit....]

In [42]:
train_df.head(13
              )

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,p kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,p kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False
5,335982,1,KKR,RCB,0,0.5,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,5,0,NaN,0,4,0,"('BB McCullum', 'SC Ganguly')",NaN,False
6,335982,1,KKR,RCB,0,0.6,bb mccullum,2,0,1,p kumar,1,1,1,0,False,legbyes,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,3,6,0,NaN,0,5,0,"('BB McCullum', 'SC Ganguly')",NaN,False
7,335982,1,KKR,RCB,1,1.1,bb mccullum,2,0,1,z khan,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,3,7,0,NaN,0,6,0,"('BB McCullum', 'SC Ganguly')",NaN,False
8,335982,1,KKR,RCB,1,1.2,bb mccullum,2,4,1,z khan,1,0,4,4,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,7,8,0,NaN,4,7,0,"('BB McCullum', 'SC Ganguly')",NaN,False
9,335982,1,KKR,RCB,1,1.3,bb mccullum,2,4,1,z khan,1,0,4,4,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,

the runs total is defines as the sumition of extra runs and runbatter

In [43]:
train_df['runs_total'].value_counts()

runs_total
1    115711
0     95778
4     32502
2     18349
6     14259
3       954
5       557
7        95
Name: count, dtype: int64

In [44]:
train_df[['runs_batter','runs_extras','runs_total']].tail(30)

,runs_batter,runs_extras,runs_total
278175,1,0,1
278176,6,0,6
278177,0,1,1
278178,0,1,1
278179,1,0,1
278180,6,0,6
278181,0,0,0
278182,0,0,0
278183,6,0,6
278184,0,0,0


In [45]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,p kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,p kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


the runs_total will help us decid the overall runs of that match 

In [46]:
train_df.groupby(['match_id', 'batting_team'])['runs_total'].sum()

match_id  batting_team
335982    KKR             222
          RCB              82
335983    CSK             240
          KXP             207
335984    DC              132
          RR              129
335985    MI              165
          RCB             166
335986    DEC             110
          KKR             112
335987    KXP             166
          RR              168
335988    DC              143
          DEC             142
335989    CSK             208
          MI              202
335990    DEC             214
          RR              217
335991    KXP             182
          MI              116
335992    RCB             135
          RR              138
335993    CSK             152
          KKR             147
335994    DEC             155
          MI              154
335995    DC              158
          KXP             162
335996    CSK             178
          RCB             165
335997    KKR             137
          MI              138
335998    DC     

In [47]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,p kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,p kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


runs_bowler

In [48]:
train_df['runs_bowler'].nunique()

8

In [49]:
train_df['runs_bowler'].unique()

array([0, 1, 4, 6, 5, 2, 3, 7])

who are the top 10  bowler with most runs in ipl history 

most expensive overs in an ipl match 4 overs 

In [50]:
train_df.groupby(['match_id','bowler'])['runs_bowler'].sum().sort_values(ascending=False).head(10)

match_id  bowler          
1473439   jc archer           76
1473464   mohammed shami      75
1473507   w o'rourke          74
1426278   mm sharma           73
1136611   basil thampi        70
1359487   yash dayal          69
1426268   rjw topley          68
1426281   l wood              68
1422126   kt maphaka          66
1178423   mujeeb ur rahman    66
Name: runs_bowler, dtype: int64

In [51]:
train_df.head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,335982,1,KKR,RCB,0,0.1,sc ganguly,1,0,1,p kumar,1,1,1,0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,335982,1,KKR,RCB,0,0.2,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,0,p kumar,0,1,1,1,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,335982,1,KKR,RCB,0,0.3,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,335982,1,KKR,RCB,0,0.4,bb mccullum,2,0,1,p kumar,1,0,0,0,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


runs_not boundry 

In [52]:
train_df['runs_not_boundary'].unique()

array([False,  True])

In [53]:
train_df['runs_not_boundary'].value_counts()

runs_not_boundary
False    278177
True         28
Name: count, dtype: int64

ther is no error in this perticular section as it define the runs made by running and not by hitting the boundary 

In [54]:
train_df[train_df['runs_not_boundary']==True].head()

,match_id,innings,batting_team,bowling_team,over,ball_no,batter,bat_pos,runs_batter,balls_faced,bowler,valid_ball,runs_extras,runs_total,runs_bowler,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
7415,336014,1,KKR,DEC,19,19.6,lr shukla,6,6,1,rp singh,1,0,6,6,True,NaN,DJ Hussey,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,SC Ganguly,Kolkata Knight Riders,23 runs,Kolkata Knight Riders,bat,"Rajiv Gandhi International Stadium, Uppal",Hyderabad,11,5,2008,2007/08,male,club,NaN,NaN,NaN,6,20,32,Unknown,Unknown,204,120,4,NaN,14,3,0,"('DJ Hussey', 'LR Shukla')",NaN,False
23804,392226,2,KXP,DC,1,1.3,sm katich,2,4,1,pj sangwan,1,0,4,4,True,NaN,S Sohal,1,NaN,NaN,NaN,121.0,NaN,NaN,NaN,NaN,False,B Lee,Kings XI Punjab,6 wickets,Kings XI Punjab,field,OUTsurance Oval,Bloemfontein,15,5,2009,2009,male,club,NaN,NaN,NaN,6,20,46,Unknown,Unknown,10,9,0,NaN,9,4,0,"('S Sohal', 'SM Katich')",NaN,False
47910,501225,2,RR,KTK,4,4.4,r dravid,2,4,1,r vinay kumar,1,0,4,4,True,NaN,SR Watson,1,NaN,NaN,NaN,110.0,NaN,NaN,NaN,NaN,False,SK Warne,Rajasthan Royals,8 wickets,Rajasthan Royals,field,Sawai Mansingh Stadium,Jaipur,24,4,2011,2011,male,club,NaN,NaN,NaN,6,20,28,Unknown,Unknown,32,28,0,NaN,18,16,0,"('R Dravid', 'SR Watson')",NaN,False
54289,501252,2,RCB,RR,5,5.1,ch gayle,1,4,1,j botha,1,0,4,4,True,NaN,TM Dilshan,2,NaN,NaN,NaN,147.0,NaN,NaN,NaN,NaN,False,S Aravind,Royal Challengers Bangalore,9 wickets,Royal Challengers Bangalore,field,Sawai Mansingh Stadium,Jaipur,11,5,2011,2011,male,club,NaN,NaN,NaN,6,20,55,Unknown,Unknown,54,31,0,NaN,21,13,0,"('CH Gayle', 'TM Dilshan')",NaN,False
55278,501257,1,KXP,DC,2,2.3,pc valthaty,1,4,1,ik pathan,1,0,4,4,True,NaN,AC Gilchrist,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,PP Chawla,Kings XI Punjab,29 runs,Delhi Daredevils,field,Himachal Pradesh Cricket Association Stadium,Dharamsala,15,5,2011,2011,male,club,NaN,NaN,NaN,6,20,60,Unknown,Unknown,11,15,0,NaN,5,8,0,"('AC Gilchrist', 'PC Valthaty')",NaN,False


In [55]:
train_df.iloc[:,15:].head()

,runs_not_boundary,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,False,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,False,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,False,NaN,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


extra runs

In [56]:
train_df['extra_type'].unique()

array(['legbyes', nan, 'wides', 'byes', 'noballs', 'penalty',
       'penalty, wides', 'byes, noballs', 'legbyes, noballs'],
      dtype=object)

In [57]:
train_df['extra_type'].value_counts()

extra_type
wides               9079
legbyes             4203
noballs             1136
byes                 686
byes, noballs         15
legbyes, noballs      12
penalty, wides         1
penalty                1
Name: count, dtype: int64

In [58]:
train_df['extra_type'].fillna('no_extra_runs',inplace=True)

C:\Users\karan\AppData\Local\Temp\ipykernel_20464\267448785.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train_df['extra_type'].fillna('no_extra_runs',inplace=True)


In [59]:
train_df['extra_type'].value_counts()

extra_type
no_extra_runs       263072
wides                 9079
legbyes               4203
noballs               1136
byes                   686
byes, noballs           15
legbyes, noballs        12
penalty, wides           1
penalty                  1
Name: count, dtype: int64

In [60]:
train_df.isnull().sum()

match_id                  0
innings                   0
batting_team              0
bowling_team              0
over                      0
ball_no                   0
batter                    0
bat_pos                   0
runs_batter               0
balls_faced               0
bowler                    0
valid_ball                0
runs_extras               0
runs_total                0
runs_bowler               0
runs_not_boundary         0
extra_type                0
non_striker               0
non_striker_pos           0
wicket_kind          264382
player_out           264382
fielders             268192
runs_target          144302
review_batter        277333
team_reviewed        277333
review_decision      277333
umpire               277333
umpires_call              0
player_of_match           0
match_won_by              0
win_outcome            4702
toss_winner               0
toss_decision             0
venue                     0
city                      0
day                 

In [61]:
train_df.iloc[:,16:].head()

,extra_type,non_striker,non_striker_pos,wicket_kind,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,legbyes,BB McCullum,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,no_extra_runs,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,wides,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,no_extra_runs,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,no_extra_runs,SC Ganguly,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


In [62]:
train_df['non_striker'].nunique()

692

In [63]:
train_df['non_striker']=train_df['non_striker'].apply(lambda x:x.lower())

In [64]:
train_df['non_striker_pos'].nunique()

11

In [65]:
train_df['non_striker_pos'].unique()

array([ 2,  1,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [66]:
train_df['wicket_kind'].nunique()

10

In [67]:
train_df['wicket_kind'].unique()

array([nan, 'caught', 'bowled', 'run out', 'lbw', 'retired hurt',
       'stumped', 'caught and bowled', 'hit wicket',
       'obstructing the field', 'retired out'], dtype=object)

In [68]:
train_df['wicket_kind'].value_counts()

wicket_kind
caught                   8665
bowled                   2345
run out                  1153
lbw                       853
caught and bowled         388
stumped                   376
hit wicket                 18
retired hurt               17
retired out                 5
obstructing the field       3
Name: count, dtype: int64

In [69]:
train_df['wicket_kind'].fillna("not out",inplace=True)

C:\Users\karan\AppData\Local\Temp\ipykernel_20464\1949907871.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train_df['wicket_kind'].fillna("not out",inplace=True)


In [70]:
train_df['wicket_kind'].value_counts()

wicket_kind
not out                  264382
caught                     8665
bowled                     2345
run out                    1153
lbw                         853
caught and bowled           388
stumped                     376
hit wicket                   18
retired hurt                 17
retired out                   5
obstructing the field         3
Name: count, dtype: int64

In [71]:
train_df.iloc[:,20:].head()

,player_out,fielders,runs_target,review_batter,team_reviewed,review_decision,umpire,umpires_call,player_of_match,match_won_by,win_outcome,toss_winner,toss_decision,venue,city,day,month,year,season,gender,team_type,superover_winner,result_type,method,balls_per_over,overs,event_match_no,stage,match_number,team_runs,team_balls,team_wicket,new_batter,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,1,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,1,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,2,0,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,3,0,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,BB McCullum,Kolkata Knight Riders,140 runs,Royal Challengers Bangalore,field,M Chinnaswamy Stadium,Bangalore,18,4,2008,2007/08,male,club,NaN,NaN,NaN,6,20,1,Unknown,Unknown,2,4,0,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


In [72]:
train_df['player_out'].value_counts()

player_out
RG Sharma                  237
V Kohli                    219
S Dhawan                   193
KD Karthik                 184
RV Uthappa                 180
SK Raina                   171
AM Rahane                  165
DA Warner                  164
AT Rayudu                  155
SV Samson                  153
MS Dhoni                   142
G Gambhir                  136
F du Plessis               136
MK Pandey                  135
AB de Villiers             130
PA Patel                   126
SR Watson                  126
CH Gayle                   126
SA Yadav                   124
KA Pollard                 121
WP Saha                    121
MA Agarwal                 120
GJ Maxwell                 118
RA Jadeja                  117
KL Rahul                   115
Yuvraj Singh               111
SS Iyer                    110
YK Pathan                  110
Q de Kock                  109
RR Pant                    104
JC Buttler                 104
BB McCullum                1